redbox-review/redbox.go

```diff
@@ -0,0 +1,293 @@
+	}
+	return rb.s3Box.Pack(row)
+}
+
+// Ship ships written data to the destination Redshift table.
+// While shipping is in progress, no other operations are permitted.
+// Ship is transactional, meaning that any returned error means
+// the destination table has remained unchanged.
+func (rb *Redbox) Ship() ([]string, error) {
+	if rb.isShipped() {
+		return nil, errBoxShipped
+	}
+	if rb.isShippingInProgress() {
+		return nil, errShippingInProgress
+	}
+
+	// Kick off the s3-to-Redshift job
+	rb.setShippingInProgress(true)
+	defer func() {
+		rb.setShippingInProgress(false)
+	}()
+
+	manifests, err := rb.s3Box.CreateManifests(rb.manifestSlug(), rb.nManifests)
+	if err != nil {
+		return nil, err
+	}
+	if len(manifests) == 0 { // If no data was written, there's nothing to ship.
+		return nil, errNothingToShip
+	}
+
+	if err := rb.copyToRedshift(manifests); err != nil {
+		return nil, err
+	}
+
+	rb.markShipped()
+	return manifests, nil
+}
+
+// manifestSlug defines a convention for the slug of each manifest file.
+func (rb *Redbox) manifestSlug() string {
+	return fmt.Sprintf("%s_%s_%s", rb.schema, rb.table, time.Now().Format(time.RFC3339))
+}
+
+// copyToRedshift transports data pointed to by the manifests into Redshift.
+// If the truncate flag is present the destination table is first cleared.
+func (rb *Redbox) copyToRedshift(manifests []string) error {
+	tx, err := rb.redshift.Begin()
+	if err != nil {
+		return err
+	}
+
+	if rb.truncate {
+		delStmt := fmt.Sprintf("DELETE FROM \"%s\".\"%s\"", rb.schema, rb.table)
+		if _, err := tx.Exec(delStmt); err != nil {
+			tx.Rollback()
+			return err
+		}
+	}
+
+	for _, manifest := range manifests {
+		copyStmt := rb.copyStatement(manifest)
+		if _, err := tx.Exec(copyStmt); err != nil {
+			tx.Rollback()
+			return err
+		}
+	}
+
+	return tx.Commit()
+}
+
+// copyStatment generates the COPY statement for the given manifest and Redbox configuration
+func (rb *Redbox) copyStatement(manifest string) string {
+	manifestURL := fmt.Sprintf("s3://%s/%s", rb.s3Bucket, manifest)
+	copy := fmt.Sprintf("COPY \"%s\".\"%s\" FROM '%s' MANIFEST REGION '%s'", rb.schema, rb.table, manifestURL, rb.s3Region)
+	dataFormat := "GZIP JSON 'auto'"
+	options := "TIMEFORMAT 'auto' TRUNCATECOLUMNS STATUPDATE ON COMPUPDATE ON"
+	creds := fmt.Sprintf("CREDENTIALS 'aws_access_key_id=%s;aws_secret_access_key=%s'", rb.awsKey, rb.awsPassword)
+	return fmt.Sprintf("%s %s %s %s", copy, dataFormat, options, creds)
+}
+
+func (rb *Redbox) setShippingInProgress(inProgress bool) {
+	rb.mt.Lock()
+	defer rb.mt.Unlock()
+	rb.shippingInProgress = inProgress
+}
+
+func (rb *Redbox) markShipped() {
+	rb.mt.Lock()
+	defer rb.mt.Unlock()
+	rb.shipped = true
+}
+
+// isShippingInProgress exposes whether a send is in progress.
+func (rb *Redbox) isShippingInProgress() bool {
+	rb.mt.Lock()
+	defer rb.mt.Unlock()
+	return rb.shippingInProgress
+}
+
+// isShipped exposes whether the box has been shipped
+func (rb *Redbox) isShipped() bool {
```

Comment:
nit: go style best practice is to not prepend `is` to functions returning booleans